Add support for bigarrays

tbrk opened the pull request:
Hello @xavierleroy,

I'm working on updating the Sundials/ML library. The new version requires `Mpi.allreduce` on "nvectors" which are interfaced using bigarrays, so I would like to add an `allreduce_bigarray` operation to ocamlmpi. This pull request is a first proposal. I added bigarray versions of other operations for good measure.

The only potential drawback of this addition is that it requires adding `bigarray.cm(x)a` to the command-line when compiling applications under older versions of OCaml. This does not seem to be required for newer OCaml versions (tested on 4.10.0).

The `(all)reduce_bigarray` calls take an `intop` even when the bigarray argument contains a floating-point or complex value and there is a dynamic check to rule out invalid operations. This is a bit ugly, but maybe using a GADT here would be overkill?

Best regards,

Tim.

tbrk committed:
Add support for bigarrays

tbrk committed:
Transparent support for MPI_IN_PLACE

Using the same bigarray as both send and receive buffer in calls to
reduce_bigarray, allreduce_bigarray, and scan_bigarray fails (at least
with openmpi 3.1.3) with

Fatal error: exception Mpi.Error("MPI_ERR_BUFFER: invalid buffer pointer")

This patch checks the buffers and substitutes MPI_IN_PLACE for the send
buffer if necessary. It has the advantage that changes to the function
interfaces are not required; one simply passes the same argument twice.

xavierleroy reviewed (commented):
Thanks for this contribution.  MPI and Bigarrays both have Fortran roots, so it makes sense to combine them!

Concerning the need to link the bigarray library: this can be made transparent to users who go through ocamlfind, as recommended.  Could you please update the META file to add a dependency on the bigarray library?

Your implementation handles one-dimension bigarrays only.  I'm OK with leaving the other bigarrays (Array2, Array3, and Genarray) for future work.  Maybe you want to name the OCaml functions with a `bigarray1` suffix, e.g. `send_bigarray1`, in preparation for a future `send_bigarray2`?  Or maybe not.

Also, I suspect the C stubs could be made to work with arbitrary-dimension bigarrays.  The main change would be in the computation of the length (number of elements).

As we discussed recently, I'm OK with the type of `reduce_bigarray`, even though some clever use of GADTs might enable a better type, including for `reduce` over normal arrays.

If you can get the META file right, I can merge as is.  The other possible changes can be left for future work.

tbrk committed:
Add a function for raising Mpi.Error

tbrk committed:
Generalize bigarray treatment

Implement the suggestion of @xavierleroy to generalize the C stubs for
bigarrays of arbitrary dimensions, and support bigarray{0,2,3}.

tbrk committed:
Add opam dependency for base-bigarray

tbrk committed:
Synchronize opam file with opam-repository

Adopt most of the distributed opam file:
https://github.com/ocaml/opam-repository/blob/master/packages/mpi/mpi.1.03/opam

tbrk commented:
The generalization to bigarrays of arbitrary dimension works quite nicely. I think I've finally understood `Bigarray.Genarray` (a bit embarrassing after all this time!).

So far, I only have two ideas for the operations signature.

1. Heavy-handed
```
type _ op =
  | Max  : 'a op
  | Min  : 'a op
  | Sum  : 'a op
  | Prod : 'a op
  | Int_land   : int op
  | Int_lor    : int op
  | Int_xor    : int op
  | Int32_land : int32 op
  | Int32_lor  : int32 op
  | Int32_xor  : int32 op
  | Int64_land : int64 op
  | Int64_lor  : int64 op
  | Int64_xor  : int64 op
  | Nativeint_land : nativeint op
  | Nativeint_lor  : nativeint op
  | Nativeint_xor  : nativeint op
  | Char_land      : char op
  | Char_lor       : char op
  | Char_xor       : char op
```

2. Bureaucratic
```
type 'a op

val op_max  : 'a op
val op_min  : 'a op
val op_sum  : 'a op
val op_prod : 'a op

module Int : sig
  val op_land : int op
  val op_lor  : int op
  val op_xor  : int op
end

module Char : sig
  val op_land : char op
  val op_lor  : char op
  val op_xor  : char op
end

module Nativeint : sig
  val op_land : nativeint op
  val op_lor  : nativeint op
  val op_xor  : nativeint op
end

module Int32 : sig
  val op_land : int32 op
  val op_lor  : int32 op
  val op_xor  : int32 op
end

module Int64 : sig
  val op_land : int64 op
  val op_lor  : int64 op
  val op_xor  : int64 op
end
```
with the obvious implementation
```
type 'a intop =
  Op_max | Op_min | Op_sum | Op_prod | Op_land | Op_lor | Op_xor

let op_max = Op_max
...
module Int = struct
  let op_land = Op_land
  ...
end
...
```

In both cases one can write:
```
val reduce_int: int -> int op -> rank -> communicator -> int
val reduce_float: float -> float op -> rank -> communicator -> float
val reduce_bigarray: ('a, 'b, 'c) Bigarray.Genarray.t  -> ('a, 'b, 'c) Bigarray.Genarray.t -> 'a op -> rank -> communicator -> unit
...
```

The second (bureaucratic) style is maybe not too bad. Is there a smarter way?

I probably should add the `Alltoall` operation too while I'm at it.

tbrk commented:
A slight variation: _bureaucratic+_
```
type 'a op

module type ALL_OPS = sig
  type t
  val op_max  : t op
  val op_min  : t op
  val op_sum  : t op
  val op_prod : t op
end

module type INT_OPS = sig
  include ALL_OPS
  val op_land : t op
  val op_lor  : t op
  val op_xor  : t op
end

module type Float = ALL_OPS with type t = float
module type Complex = ALL_OPS with type t = Complex.t
module type Int32 = INT_OPS with type t = int32
module type Int64 = INT_OPS with type t = int64
module type Nativeint = INT_OPS with type t = nativeint
module type Char = INT_OPS with type t = char
```

tbrk committed:
Add support for MPI_Alltoall

The generic version fuses the existing `scatter` and `gather` implementations.

tbrk commented:
A more sophisticated attempt:
```
type _ op =
  | Max  : [< `Int of int | `Char of char | `Float of float ] op
  | Land : [< `Int of int | `Char of char ] op

module type S = sig
  val reduce_int: int -> [`Int of int ] op -> int
  val reduce_float: float -> [`Float of float] op -> float
  val reduce_list: 'a list -> [< `Int of 'a | `Char of 'a | `Float of 'a ] op -> unit
end

module A : S = struct
  let reduce_int x op = x
  let reduce_float x op = x
  let reduce_list s op = ()
end
```

It works well for the pre-bigarray interface:
```
let _ = A.reduce_int 0 Max       (* OK *)
let _ = A.reduce_int 0 Land      (* OK *)
let _ = A.reduce_float 0. Max    (* OK *)
let _ = A.reduce_float 0. Land   (* OK: REJECTED *)
```

But, unfortunately, I haven't found a way to constrain the operation type based on a type variable in another argument.
```
let _ = A.reduce_list [0] Max    (* OK *)
let _ = A.reduce_list [0] Land   (* OK *)
let _ = A.reduce_list [0.] Max   (* OK *)
let _ = A.reduce_list [0.] Land  (* KO: NOT REJECTED! *)
```

The type inferred for `A.reduce_list [0.]` is
```
_[< `Char of float | `Float of float | `Int of float ] op -> unit
```
which I would have thought perfect, but there are obviously many things that I don't understand about the typing of polymorphic variables and GADTs!

tbrk committed:
Single GADT for operations

Result of discussions with @xavierleroy and (online)[1].

- A single type for reduce and scan operations.
- The use of GADT prevents passing integer-only operations to float-only
  functions like `reduce_float` and `reduce_float_array`.
- A dynamic check is used for operations on bigarrays.
- The previous constructors are maintained for backward compatibility
  but have been marked as deprecated.

[1]: https://discuss.ocaml.org/t/filtering-constructors-by-type-variable-gadts-polymorphic-variants/6195

tbrk committed:
Bump minimum OCaml version to 4.06.0

It turns out that `caml_ba_num_elts` is only exposed in 4.06.0 <= OCaml.

tbrk committed:
Remove the OCaml-level verification on operations

This verification fails type checking in OCaml < 4.08.0.

It is unnecessary because MPI already checks that an operation is valid.
Trying to pass an integer-only operation for a reduce or scan of a
bigarray containing floating-point values fails with
`exception Mpi.Error("MPI_ERR_OP: invalid reduce operation")`.

tbrk committed:
Fix tests for OCaml 4.06.0

tbrk commented:
Here's a tweaking of the operations per our previous discussion. I did not adopt the suggestions on [OCaml discuss](https://discuss.ocaml.org/t/filtering-constructors-by-type-variable-gadts-polymorphic-variants/6195). They are clever and remove the need for a dynamic check, but they require that the caller pass a witness for the type.

The generalization of the bigarray interface uses `caml_ba_num_elts`. It turns out that this function was only exported by the OCaml runtime in 4.06.0. I could duplicate this function locally if you prefer to maintain compatibility for 4.03.0 <= OCaml < 4.06.0.

xavierleroy committed:
Documentation updates

- Don't deprecate the [Int_*] and [Float_*] operations right now.
  Users need a couple of releases to get used to the new operation names.
- Use [Land] [Lor] [Xor] instead of [Int_land] [Int_lor] [Int_xor]
  in the documentation of the reduce functions.

xavierleroy reviewed (approved):
Looks good to me!  I pushed directly a couple of updates to the `mpi.mli` interface.  The main change is that we shouldn't mark something `[@deprecated]` until the more modern replacement has been available for a few releases.  Otherwise users can't write warning-free code that compiles with several versions of the library.

Let me know if you're OK with the documentation updates and I'll merge.

tbrk commented:
Looks good to me! Thank you.
